falcon/util/structures.py

```diff
@@ -107,6 +107,126 @@ def __repr__(self):
         return '%s(%r)' % (self.__class__.__name__, dict(self.items()))
 
 
+class Context(object):
```

Comment:
It may be helpful to add an inline `# NOTE` comment to explain why this does not inherit from `MutableMapping`. I assume it is (1) to avoid breaking any app code later on that may rely on checking to see if `Context` subclasses `MutableMapping` if we remove the mapping interface, and (2) to avoid an unnecessary inheritance chain, since for the sake of efficiency all of the default implementations provided by the ABC have to be overridden.

Reply:
It may also be helpful to add a `# PERF` comment explaining the overhead introduced by this implementation vs. using a `dict` directly. From a quick test on my MBP, it looks like the extra level of indirection adds about 200 ns, so probably not enough for anyone to notice.

Reply:
Tbh the concern regarding `MutableMapping` is a good question, I remember I was looking at it briefly, and made a decision not to inherit from it, but haven't documented why. I'll recheck if there is any performance penalty associated to `MutableMapping`. If not anything else, yes (1), I didn't want to cement the fact that it is implementing the mapping interface as I wanted this to be viewed more as a bare class with convenience methods to ease migration.